src/NHibernate/Impl/SessionFactoryImpl.cs

```diff
@@ -1281,9 +1281,15 @@ private ICurrentSessionContext BuildCurrentSessionContext()
 
 			try
 			{
-				System.Type implClass = ReflectHelper.ClassForName(impl);
-				return
-					(ICurrentSessionContext)Environment.ObjectsFactory.CreateInstance(implClass, new object[] { this });
+				var implClass = ReflectHelper.ClassForName(impl);
+				if (!typeof(ISessionFactoryAwareCurrentSessionContext).IsAssignableFrom(implClass))
+				{
+					return (ICurrentSessionContext) Activator.CreateInstance(implClass, this);
+				}
+
+				var context = (ISessionFactoryAwareCurrentSessionContext) Environment.ObjectsFactory.CreateInstance(implClass);
+				context.SetFactory(this);
+				return context;
```

Comment:
I was thinking a lot about this chunk of code, but could not come up with a great solution, unfortunately.

So, I came up with this:

```csharp

var implClass = ReflectHelper.ClassForName(impl);
var constructor = implClass.GetConstructor(new [] { typeof(ISessionFactoryImplementor) });
ICurrentSessionContext context;
if (constructor != null)
{
  context = (ICurrentSessionContext) constructor.Invoke(new object[] { this });
}
else
{
  context = (ICurrentSessionContext) Environment.ObjectsFactory.CreateInstance(implClass);
}
if (context is ISessionFactoryAwareCurrentSessionContext sessionFactoryAwareContext)
{
  sessionFactoryAwareContext.SetFactory(this);
}
```

Reply:
Ok, changed for this. (And rebased.)